Libraries/LibWeb/Streams/AbstractOperations.cpp

```diff
@@ -289,8 +289,372 @@ bool readable_stream_has_default_reader(ReadableStream const& stream)
+                GC::create_function(heap(), [this](JS::Value new_error) -> WebIDL::ExceptionOr<JS::Value> {
+                    finish(new_error);
+                    return JS::js_undefined();
+                }));
+        };
+
+        // 3. If dest.[[state]] is "writable" and ! WritableStreamCloseQueuedOrInFlight(dest) is false,
+        if (m_destination->state() == WritableStream::State::Writable && !writable_stream_close_queued_or_in_flight(m_destination)) {
+            // 1. If any chunks have been read but not yet written, write them to dest.
+            write_unwritten_chunks();
+
+            // 2. Wait until every chunk that has been read has been written (i.e. the corresponding promises have settled).
+            wait_for_pending_writes_to_complete(move(on_pending_writes_complete));
+        } else {
+            on_pending_writes_complete();
+        }
+    }
+
+    // https://streams.spec.whatwg.org/#rs-pipeTo-shutdown
+    void shutdown(Optional<JS::Value> error = {})
+    {
+        // 1. If shuttingDown is true, abort these substeps.
+        if (m_shutting_down)
+            return;
+
+        // 2. Set shuttingDown to true.
+        m_shutting_down = true;
+
+        auto on_pending_writes_complete = [this, error = move(error)]() mutable {
+            HTML::TemporaryExecutionContext execution_context { m_realm, HTML::TemporaryExecutionContext::CallbacksEnabled::Yes };
+
+            // 4. Finalize, passing along error if it was given.
+            finish(move(error));
+        };
+
+        // 3. If dest.[[state]] is "writable" and ! WritableStreamCloseQueuedOrInFlight(dest) is false,
+        if (m_destination->state() == WritableStream::State::Writable && !writable_stream_close_queued_or_in_flight(m_destination)) {
+            // 1. If any chunks have been read but not yet written, write them to dest.
+            write_unwritten_chunks();
+
+            // 2. Wait until every chunk that has been read has been written (i.e. the corresponding promises have settled).
+            wait_for_pending_writes_to_complete(move(on_pending_writes_complete));
+        } else {
+            on_pending_writes_complete();
+        }
+    }
+
+private:
+    ReadableStreamPipeTo(
+        GC::Ref<JS::Realm> realm,
+        GC::Ref<WebIDL::Promise> promise,
+        GC::Ref<ReadableStream> source,
+        GC::Ref<WritableStream> destination,
+        GC::Ref<ReadableStreamDefaultReader> reader,
+        GC::Ref<WritableStreamDefaultWriter> writer,
+        bool prevent_close,
+        bool prevent_abort,
+        bool prevent_cancel)
+        : m_realm(realm)
+        , m_promise(promise)
+        , m_source(source)
+        , m_destination(destination)
+        , m_reader(reader)
+        , m_writer(writer)
+        , m_prevent_close(prevent_close)
+        , m_prevent_abort(prevent_abort)
+        , m_prevent_cancel(prevent_cancel)
+    {
+        m_reader->set_readable_stream_pipe_to_operation({}, this);
+    }
```

Comment:
This formatting sure is exciting. I'm not sure I'm a fan, it makes it hard to know where the parameters end and member vars begin :thinking:

Reply:
eh, my eyes go to the lines starting with `:` or `,` to see the members. It was just all too much to be horizontal.